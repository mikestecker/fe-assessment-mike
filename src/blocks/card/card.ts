import { Card } from "../../types/Card";
import data from "../card/card.json";
import "../card/card.scss";

/**
 * Card decorator function
 *
 * This function accepts props defined in types/Card.ts and
 * marshals in data from the card.json file located in this
 * file's directory.
 *
 * From here, we would look to target the "app" div defined
 * in our "card.html" file where we will now be able to append
 * a parent (wrapper) element and using all the child elements
 * this component comprises to create a re-usable component.
 *
 * @param {string} title Title text (h1)
 * @param {string} body Body text (p)
 * @param {Object} image Image source/alt
 * @param {Object} button Button href/label
 */
function decorate({ title, body, image, button }: Card) {
  const app = document.getElementById("app");
  const parent = document.createElement("div");

  // Image (if defined)
  if (image) {
    const imgEl = document.createElement("img");
    imgEl.classList.add("card-img");
    imgEl.src = image.src;
    imgEl.alt = image.alt;
    parent.append(imgEl);
  }

  // Title
  const titleEl = document.createElement("h1");
  titleEl.classList.add("card-title");
  titleEl.textContent = title;
  parent.append(titleEl);

  // Body
  const contentEl = document.createElement("p");
  contentEl.classList.add("card-body");
  contentEl.textContent = body;
  parent.append(contentEl);

  // Button/Link (if defined)
  if (button) {
    const btnEl = document.createElement("a");
    btnEl.classList.add("card-btn");
    btnEl.href = button.href;
    btnEl.title = button.label;
    btnEl.textContent = button.label;
    btnEl.classList.add("btn");
    parent.append(btnEl);
  }

  parent.classList.add("card");

  app?.appendChild(parent);
}

decorate(data);
