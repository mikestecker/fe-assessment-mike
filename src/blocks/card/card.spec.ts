import * as fs from "fs";
import * as path from "path";
const html = fs.readFileSync(path.resolve(__dirname, "./card.html"));

jest.dontMock("fs");

describe("Card ->", function () {
  beforeEach(function () {
    document.documentElement.innerHTML = html.toString();
    require("./card");
  });

  afterEach(function () {
    jest.resetModules();
  });

  it("should have access to DOM", function () {
    expect(document.getElementById("app")).toBeTruthy();
  });

  it("should contain exactly one div with class of .card", function () {
    const cards = document.getElementsByClassName("card");
    expect(cards.length).toBe(1);
  });

  it("should match snapshot", () => {
    const card = document.getElementsByClassName("card")[0];
    expect(card).toMatchSnapshot();
  });

  it("should contain a .card-title inside the .card wrapper", () => {
    const card = document.getElementsByClassName("card")[0];
    expect(card.getElementsByClassName("card-title")).toBeTruthy();
  });

  it("should contain a .card-content inside the .card wrapper", () => {
    const card = document.getElementsByClassName("card")[0];
    expect(card.getElementsByClassName("card-content")).toBeTruthy();
  });
});
