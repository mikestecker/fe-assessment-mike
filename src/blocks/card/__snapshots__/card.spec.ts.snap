// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card -> should match snapshot 1`] = `
<div
  class="card"
>
  <img
    alt="powerlines"
    class="card-img"
    src="https://www.wesco.com/content/dam/wesco/assets/wesco/images/600/Power_Lines_Orange_Sunset_1316185179_RGB.jpg"
  />
  <h1
    class="card-title"
  >
    Powerlines
  </h1>
  <p
    class="card-body"
  >
    Here's some awesome descriptive text about powerlines! The point of this text is to test responsive styling.
  </p>
  <a
    class="card-btn btn"
    href="#"
    title="Learn more"
  >
    Learn more
  </a>
</div>
`;
